cranelift/codegen/src/isa/s390x/lower.rs

```diff
@@ -2565,17 +2578,26 @@ fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
 
         Opcode::Trapif => {
             let condcode = ctx.data(insn).cond_code().unwrap();
-            let cond = Cond::from_intcc(condcode);
+            let mut cond = Cond::from_intcc(condcode);
             let is_signed = condcode_is_signed(condcode);
 
-            // Verification ensures that the input is always a single-def ifcmp.
             let cmp_insn = ctx
                 .get_input_as_source_or_const(inputs[0].insn, inputs[0].input)
                 .inst
                 .unwrap()
                 .0;
-            debug_assert_eq!(ctx.data(cmp_insn).opcode(), Opcode::Ifcmp);
-            lower_icmp_to_flags(ctx, cmp_insn, is_signed, true);
+            if ctx.data(cmp_insn).opcode() == Opcode::IaddIfcout {
+                // The flags must not have been clobbered by any other instruction between the
+                // iadd_ifcout and this instruction, as verified by the CLIF validator; so we
+                // can simply rely on the condition code here.  However, the encoding is
+                // different.
+                assert!(condcode == IntCC::UnsignedGreaterThan);
+                cond = Cond::from_mask(2 | 1);
```

Comment:
Could we add a comment here noting what these constants are (which condition flags) and how the combination of the add instruction generated by `IaddIfcout` and the condcode here will properly trap?

Perhaps also a note in `unsigned_add_overflow_condition()` would be helpful in order to describe how the condition code is actually specially handled in the one place this is known to be used.